### PR TITLE
#163518691 Comment on an article

### PR DESCRIPTION
**Pull request by Musigwa-zz**
#### What does this PR do?
- It helps the logged in user to comment on any of the articles
#### Description of Task to be completed?
- Write the tests for the comment on an article
- Secure the endpoint with JsonWebToken
- Create the article

#### How should this be manually tested?
- After cloning the repository, check out `ft-create-article-endpoints-163518689` branch
- Make sure to read `README.md` file to get the setup instructions
- Using any `API client tool(usually POSTMAN)`,
- Create the user and login to get the `access-token`
- Create an article using `POST /api/v1/articles` endpoint
- Comment on the created article using `POST /api/v1/articles/:slug/comments` endpoint
- Make sure to pass the token under `request headers`. It should be {`Authorization`: `Bearer <access-token>`} KEY, VALUE PAIR
- The request's body should be as mentioned in the `README.md` file under `Endpoints:` ==> `Add Comments to an Article` subtitle
After the successful article creation, you should get back the article information. with the successful message.

#### Any background context you want to provide?
All tests related to the "reset_password" are failing because of the `Maximum credits exceeded`
#### What are the relevant pivotal tracker stories?
- #163518691

### CODE REVIEW

**Review by YvesIraguha — APPROVED** [review]

**Review by abayo-luc — CHANGES_REQUESTED** [review]
- When the user entered nonexisting slag, the request is hanging.

File: controllers/articles.js [hunk 1]
```diff
@@ -0,0 +1,53 @@
+import models from "../models";
+import constants from "../helpers/constants";
+import articleValidator from "../helpers/validators/article";
+
+const { User, Article, Tag, ArticleTag } = models;
+const { CREATED, NOT_FOUND, BAD_REQUEST } = constants.statusCode;
+
+export default class ArticleController {
```
Comment (abayo-luc) [inline controllers/articles.js]: Waaahooo, a very nice clean code. Keep this up!!

File: controllers/articles.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+              status: CREATED,
+              message: "Article created",
+              article: { ...article.dataValues, tagsList }
+            })
+          : res
+              .status(NOT_FOUND)
+              .json({ status: NOT_FOUND, message: "Please consider logging in" });
+      }
+    } catch (error) {
+      // console.log("DATABASE ERROR:", error);
```
Comment (abayo-luc) [inline controllers/articles.js]: remove this

File: tests/integration/comment.test.js [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+import { newArticle, newUser, newComment } from "../testData";
+import constants from "../../helpers/constants";
+
+let token;
+let slug;
+
+const { UNAUTHORIZED, CREATED, BAD_REQUEST } = constants.statusCode;
+chai.use(chaiHttp);
+
+before(done => {
```
Comment (abayo-luc) [inline tests/integration/comment.test.js:13]: I can't see where you are cleaning the database after testing.

**Review by abayo-luc — DISMISSED** [review]
- When the user entered nonexisting slag, the request is hanging. 
- You have clean code, very readable. 
- Make sure you are clearing the DB after testing.

File: routes/index.js [hunk 4]
```diff
@@ -1,8 +1,14 @@
...
+import articleRoutes from "./articles";
+import checkAuth from "../middlewares/checkAuth";
+import commentRoutes from "./comments";
 
 const router = Router();
 
+// PUBLIC ROUTES
 router.use(userRouters);
+// PRIVATE(PROTECTED) ROUTES
+router.use(checkAuth, articleRoutes, commentRoutes);
```
Comment (niomwungeri-fabrice) [inline routes/index.js]: Not all endpoints related to the article requires authentication, for example: fetching one article and fetch all articles `checkAuth` should be on specific routes, not all endpoints related to the articles. @Musigwa

File: tests/integration/article.test.js [hunk 5]
```diff
@@ -0,0 +1,135 @@
+import chaiHttp from "chai-http";
+import chai, { expect, should } from "chai";
+import app from "../../index";
+import { newArticle, newUser } from "../testData";
```
Comment (niomwungeri-fabrice) [inline tests/integration/article.test.js]: we are store data for testing purpose in different places, can we have one? @Musigwa

**Review by niomwungeri-fabrice — DISMISSED** [review]

File: tests/integration/comment.test.js (hunk 3 above)
Comment (Musigwa-zz) [reply tests/integration/comment.test.js:13]: Okay, let me work on it.

File: routes/index.js (hunk 4 above)
Comment (Musigwa-zz) [reply routes/index.js]: Okay, I'll fix that

File: tests/integration/article.test.js (hunk 5 above)
Comment (Musigwa-zz) [reply tests/integration/article.test.js]: Absolutely, I prefer that we all use the file I added called: ==> testData and it should be in tests/ directory to respect Single Responsibility Principle(RSP). I think by the moment this PR will be merged, everyone should respect that.

**Review by abayo-luc — CHANGES_REQUESTED** [review]
In the future when you want to add a column to an existing table, please create a new migration. Add the column to an existing migration cost all of us to drop our DB, and you know that can't be done in production env.

File: tests/integration/comment.test.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                  done(err || undefined);
+                });
+            }
+          });
+      }
+    });
+});
+
+describe("# Article's comments endpoints", () => {
+  describe("POST /articles/:slug/comments", () => {
```
Comment (abayo-luc) [inline tests/integration/comment.test.js]: This second description is not needed, plus it has an irrelevant description.

**Review by abayo-luc — APPROVED** [review]

File: models/Article.js [hunk 7]
```diff
@@ -16,26 +18,22 @@ export default (sequelize, DataTypes) => {
...
-      //       Bcrypt.genSaltSync(process.env.SALTROUNDS),
-      //       null
-      //     );
-      //     newUser.password = hash;
-      //   }
-      // }
+      tableName: "articles",
+      hooks: {
+        beforeCreate(article) {
+          /* eslint no-bitwise: "off" */
```
Comment (puritybirir) [inline models/Article.js]: Do not disable rule

File: tests/integration/comment.test.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        .request(app)
+        .post(`/api/v1/articles/${slug}/comments`)
+        .send(newComment)
+        .end((err, res) => {
+          expect(res.status).equals(UNAUTHORIZED);
+          expect(res.body.message).to.contain("Please provide a token");
+          done();
+        });
+    });
+
```
Comment (puritybirir) [inline tests/integration/comment.test.js:77]: Remove space or add

**Review by puritybirir — CHANGES_REQUESTED** [review]

**Review by puritybirir — APPROVED** [review]

**Pull request merged**